File: 프로그래머스/lv3/43164. 여행경로/여행경로.py
from collections import defaultdict
tickets_dict = defaultdict(list)
ans = []

def dfs(now:str, _li, N):
    if len(_li) == N:
        global ans
        ans.append(_li)
        return
    for i in range(len(tickets_dict[now])):
        next, visited = tickets_dict[now][i][0], tickets_dict[now][i][1]
        if not visited:
            tickets_dict[now][i][1] = True
            dfs(next, _li + [next], N)
            tickets_dict[now][i][1] = False


def solution(tickets):

    for tik in tickets:
        tickets_dict[tik[0]].append([tik[1], False])
        tickets_dict[tik[0]].sort()
    global ans

    dfs('ICN', ['ICN'], len(tickets) + 1)

    ans.sort(key=lambda x: "".join(x))

    return ans[0]

File: 프로그래머스/lv3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    boolean[] visited;
    ArrayList<String> allRoute = new ArrayList<>();
    
    public String[] solution(String[][] tickets) {
        visited = new boolean[tickets.length];
        Arrays.fill(visited, false);
        
        dfs("ICN", "ICN", tickets, 0);
        Collections.sort(allRoute);
        return allRoute.get(0).split(" ");
    }
    public void dfs(String src, String route, String[][] tickets, int cnt){
        if(cnt == tickets.length){
            allRoute.add(route);
            return;
        }
        for(int i = 0; i < tickets.length; ++i){
            if(tickets[i][0].equals(src) && !visited[i]){
                visited[i] = true;
                dfs(tickets[i][1], route + " " + tickets[i][1], tickets, cnt + 1);
                visited[i] = false;
            }
        }
    }
}